design/open/staging-compaction.md

```diff
@@ -0,0 +1,172 @@
+atomically update the branch entity:
+
+1. Set the `CompactedMetaRangeID` to the new MetaRangeID.
+1. Remove the compacted `SealedTokens` from the branch.
+
+The Compactor should fail the operation if the sealed tokens have changed 
+since the compaction started (SetIf). Although the algorithm can tolerate 
+additional sealed tokens being added during the compaction, it's better to 
+avoid competing with concurrent commits. Commits have the same benefits as 
+compactions, but they are proactively triggered by the user (and might fail 
+if compaction succeeds).
+
+Consistency guarantees for the compaction process are derived from the 
+Commit operation consistency guarantees. In short, the compaction process 
+follows the same principles as the Commit operation. Writes to a branch 
+staging token succeeds only if it's the staging token after the write 
+occurred. Therefore, replacing the sealed token with an equivelant MetaRange 
+guarantees no successful writes are gone missing.
+
+### Branch Write Flow
+
+No change here. Writes are performed on the staging token.
+
+### Branch Read Flow
+
+Reading from a branch includes both a specific entry lookup and a listing 
+request on entries of a branch. The read flow is more complicated than what 
+we have today, and is different for compacted and uncompacted branches.
+
+#### Combined (Committed+ Uncommitted)
+
+There are 3 layers of reading from a branch:
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. Depends on compaction:
+   1. If a CompactedMetaRangeID exists, read from the compacted MetaRange.
+   1. If a CompactedMetaRangeID doesn't exist, read from the CommitID.
+
+#### Committed
+
+Just like today, reads are performed on the Branch's CommitID.
+
+#### Uncommitted
+
+For operations such as `DiffUncommitted` or checking if the staging 
+area is empty, the read flow will be as follows:
+
+1. Read from the staging token.
+2. Read from the sealed tokens (in order).
+3. If a CompactedMetaRangeID exists, read the 2-way diff between the compacted
+   metarange and the CommitID's metarange.
+
+* There's an inefficiency here, as there's an option we'll need to read 2 whole 
+  metaranges to get the diff, like when there's a single change in every 
+  range. The nature of changes to a lakeFS branch is such that changes are 
+  expected in a small number of ranges, and the diff operation is expected 
+  to skip most ranges. Moreover, Pebble caching the ranges should remove the 
+  most costly operation of ranges comparison - fetching them from S3. If 
+  this is still inefficient, we can use the immutability trait[^3] of the 
+  diff: we can calculate the diff result once and cache it.
+
+### Commit Flow
+
+The commit flow is slightly affected by the compaction process. If 
+compaction never happened, the commit flow is the same as today. If a 
+compaction happened, apply the changes to the compacted metarange instead of 
+the HEAD commit metarange.
```

Comment:
Guess we will need to address "empty commit" differently in the case of compaction metarange.

Reply:
Why are you calling out `empty commit` specifically? The view of the staging remains the same. You look at staging token + sealed tokens + compaction (if available) and see if they present any changes to the commit.

Reply:
Wait... we are overloading the word "commit" here and I think that it leads to some confusion.  AFAICT we just need to set [`AllowEmpty`](https://github.com/treeverse/lakeFS/blob/f5457058c6a9d78cefca7e7a5fe4d95c269129e3/pkg/graveler/graveler.go#L453) Let me see if I understand how compaction works.  I think we are talking about a scenario like this one: I add many (100_000) objects staging, and delete them without committing.

Now I have a staging area with 100_000 tombstones[^*].  A few seconds later I want to compact.  This is not a whole commit!  We generate a new metarange using committed.Commit -- this is below Graveler.  _We told it to AllowEmpty._  It runs over the previous base metarange (branch.CompactedMetaRange or, if we don't have, the metarange of branch.CommitID and applies all the tombstones.  _Nothing happens_, so it generates the exact same ranges and metaranges as before.  When we close the MetaRangeWriter we get the exact same metarange ID.  It _only_ deletes the staging token and the sealed tokens, the base metarange ID stays.

So AFAICS this is the same flow with no changes.  The "no empty commits" rule just means we need to set AllowEmpty here.


[^*]:  Reminder: we cannot _delete_ the key from the staging area, because the original uncommitted write might have overwritten a previous object.  We never scan backwards, and doing so on deletions would make them expensive.